sched/adaptdl_sched/allocator.py

```diff
@@ -38,27 +38,93 @@ class AdaptDLAllocator(object):
     def __init__(self, expander):
         self._core_api = kubernetes.client.CoreV1Api()
         self._objs_api = kubernetes.client.CustomObjectsApi()
+        self._custom_resource = ("adaptdl.petuum.com", "v1",
+                                 "", "adaptdljobs")
         self._cluster_expander = expander
         self._policy = PolluxPolicy()
+        # lock for the two corountines in run()
+        self.lock = asyncio.Lock()
 
     async def run(self):
+        # two functionality: (1) watch for new job and start if possible.
+        # (2) periodically optimize existing jobs
+        await asyncio.gather(
+            self._allocate_one_loop(),
+            self._optimize_all_loop()
+        )
+
+    async def _allocate_one_loop(self):
+        async with kubernetes.watch.Watch() as watch:
+            while True:
+                async for event in watch.stream(
+                        self._objs_api.list_namespaced_custom_object,
+                        *self._custom_resource, timeout_seconds=60):
+                    if event["type"] == "ADDED":  # there is a n arriving job
+                        async with self.lock:
+                            await self._allocate_one(event)
+
+    async def _allocate_one(self, event):
+        # re-read the job , compared with the previous readjob
+        job = event["object"]
+        namespace = job["metadata"]["namespace"]
+        name = job["metadata"]["name"]
+
+        job = await self._objs_api.get_namespaced_custom_object(
+            "adaptdl.petuum.com", "v1", namespace, "adaptdljobs", name
+        )
+        # some other coroutine has handled this
+        if job.get("status", {}).get("allocation") or\
+                job.get("status", {}).get("group"):
+            return
+
+        namespace = job["metadata"]["namespace"]
+        name = job["metadata"]["name"]
+        # if this is a restarted job, skip i
+        LOG.info("detected an added job %s/%s.",
+                 namespace, name)
+        # parse the job infomation
+        job_info = {}
+        job_info[(namespace, name)] = self._get_job_info(job)
+
+        # find available nodes
+        node_info_list, _ = await self._find_nodes()
+
+        # get the node to allocate
+        new_allocation = self._policy._allocate_job(
+                            job_info, node_info_list)
```

Comment:
I think there is a bug here. The node resources contain the amount of resources not currently used by any AdaptDL job. So this will pass in more resources than are actually available, and cause the allocator to request more resources than there are available.